### PR TITLE
Fix cxx flags dir

### LINKED ISSUE #4847: foreign_library with c++ files does not add stdc++ dependency
## Expected Behavior

Be able to compile an executable in bytecode while linking with a library containing c++ foreign_archive.

To do so, the ocamlmklib should be invoked with `-lstdc++`.

<details>
<summary>Dynamique section</summary>

```
$ readelf -d _build/default/lib/dllhello_world.so 

Dynamic section at offset 0x2db0 contains 27 entries:
  Tag        Type                         Name/Value
 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
 0x000000000000000c (INIT)               0x1000
 0x000000000000000d (FINI)               0x11bc
 0x0000000000000019 (INIT_ARRAY)         0x3d98
 0x000000000000001b (INIT_ARRAYSZ)       16 (bytes)
 0x000000000000001a (FINI_ARRAY)         0x3da8
 0x000000000000001c (FINI_ARRAYSZ)       8 (bytes)
 0x000000006ffffef5 (GNU_HASH)           0x2f0
 0x0000000000000005 (STRTAB)             0x420
 0x0000000000000006 (SYMTAB)             0x318
 0x000000000000000a (STRSZ)              309 (bytes)
 0x000000000000000b (SYMENT)             24 (bytes)
 0x0000000000000003 (PLTGOT)             0x3fa0
 0x0000000000000002 (PLTRELSZ)           72 (bytes)
 0x0000000000000014 (PLTREL)             RELA
 0x0000000000000017 (JMPREL)             0x6b0
 0x0000000000000007 (RELA)               0x5c0
 0x0000000000000008 (RELASZ)             240 (bytes)
 0x0000000000000009 (RELAENT)            24 (bytes)
 0x000000000000001e (FLAGS)              BIND_NOW
 0x000000006ffffffb (FLAGS_1)            Flags: NOW
 0x000000006ffffffe (VERNEED)            0x570
 0x000000006fffffff (VERNEEDNUM)         2
 0x000000006ffffff0 (VERSYM)             0x556
 0x000000006ffffff9 (RELACOUNT)          4
 0x0000000000000000 (NULL)               0x0
```

</details>

## Actual Behavior

<details>
<summary>Dynamique section</summary>

```
$ readelf -d _build/default/lib/dllhello_world.so 

Dynamic section at offset 0x2dc0 contains 26 entries:
  Tag        Type                         Name/Value
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
 0x000000000000000c (INIT)               0x1000
 0x000000000000000d (FINI)               0x11bc
 0x0000000000000019 (INIT_ARRAY)         0x3da8
 0x000000000000001b (INIT_ARRAYSZ)       16 (bytes)
 0x000000000000001a (FINI_ARRAY)         0x3db8
 0x000000000000001c (FINI_ARRAYSZ)       8 (bytes)
 0x000000006ffffef5 (GNU_HASH)           0x2f0
 0x0000000000000005 (STRTAB)             0x420
 0x0000000000000006 (SYMTAB)             0x318
 0x000000000000000a (STRSZ)              268 (bytes)
 0x000000000000000b (SYMENT)             24 (bytes)
 0x0000000000000003 (PLTGOT)             0x3fa0
 0x0000000000000002 (PLTRELSZ)           72 (bytes)
 0x0000000000000014 (PLTREL)             RELA
 0x0000000000000017 (JMPREL)             0x658
 0x0000000000000007 (RELA)               0x568
 0x0000000000000008 (RELASZ)             240 (bytes)
 0x0000000000000009 (RELAENT)            24 (bytes)
 0x000000000000001e (FLAGS)              BIND_NOW
 0x000000006ffffffb (FLAGS_1)            Flags: NOW
 0x000000006ffffffe (VERNEED)            0x548
 0x000000006fffffff (VERNEEDNUM)         1
 0x000000006ffffff0 (VERSYM)             0x52c
 0x000000006ffffff9 (RELACOUNT)          4
 0x0000000000000000 (NULL)               0x0
```
</details>

```
$ dune build --root .
File "_none_", line 1:
Error: Error on dynamically loaded library: ./lib/dllhello_world.so: ./lib/dllhello_world.so: undefined symbol: _ZSt4cout
```

## Reproduction

- PR with a reproducing test: 

1. #4846 

## Specifications

- Version of `dune` (output of `dune --version`): 2.9
- Version of `ocaml` (output of `ocamlc --version`): 4.08
- Operating system (distribution and version): ubuntu 20.04

## Additional notes

- The `foreign_library` should support `library_flags` and `c_library_flags` in order to add other dependencies
- Would be helpful to be able to create a `foreign_library` mixing c and c++ files (e.g. `(languages c cxx)`)

### PR DESCRIPTION
**Pull request by voodoos**
Default C++ flags require compiler detection and need the path to the build directory for that. 
Previously the wrong path was sometime given. This PR fixes the issue and changed the signature of the function to require the context instead of the build path.

I discovered this problem when rebasing #4846 

CC @recoules can you confirm that it fixes #4847 ? And maybe test it in a more realist setting ?

### CODE REVIEW

**Comment by voodoos** [comment]
> CC @recoules can you confirm that it fixes #4847 ? And maybe test it in a more realist setting ?

Or maybe later... There is an issue when the compiler is  gcc apparently, I will have a look.

File: test/blackbox-tests/test-cases/foreign_cxx_library.t/run.t [hunk 1]
```diff
@@ -0,0 +1,2 @@
+Building a cxx library should run ocamlmklib with "-lstdc++".
+  $ dune build
```
Comment (jeremiedimino) [inline test/blackbox-tests/test-cases/foreign_cxx_library.t/run.t:2]: How should I check that this test succeeds?

Comment (voodoos) [reply test/blackbox-tests/test-cases/foreign_cxx_library.t/run.t:2]: If it fails a bunch of `undefined symbols` such as `std::__1::cout` errors are printed by the linker.
I will make that clear in the test description. Also the flag is not always `-stdc++`, it is `-lc++` for clang.

Also there is no specific link flag for c++ compilation with MSVC so this test won't fail on windows.

Comment (jeremiedimino) [reply test/blackbox-tests/test-cases/foreign_cxx_library.t/run.t:4]: Ack. I'm not sure we run the testsuite on Windows BTW.

Comment (jeremiedimino) [reply test/blackbox-tests/test-cases/foreign_cxx_library.t/run.t:4]: Just one last thing: I was looking at the commits individually. I was expecting to see the test fail at first and then observe subsequent commits fix it, but it seems that it was succeeding from the start. Is that the case?

Comment (voodoos) [reply test/blackbox-tests/test-cases/foreign_cxx_library.t/run.t:4]: Yes they are indeed failing but the error message was not promoted, do you want me to edit it ?

Comment (jeremiedimino) [reply test/blackbox-tests/test-cases/foreign_cxx_library.t/run.t:4]: That's alright, I just wanted to check it was the case. For next time, please do promote it. It's good practice to first demonstrate the problem and then demonstrate that it is fixed. Plus it's good if the testsuite passes for each commit.

**Review by jeremiedimino — APPROVED** [review]
Looks good!

**Comment by recoules** [comment]
> > CC @recoules can you confirm that it fixes #4847 ? And maybe test it in a more realist setting ?
> 
> ~Or maybe later... There is an issue when the compiler is gcc apparently, I will have a look.~ 
> 
> Should be fixed.

It worked for my original use case.

**Pull request merged**